lib/reports/dynamic.rb

```diff
@@ -0,0 +1,266 @@
+    end
+
+    # Puts it all together.
+    def run
+      outf header
+      records do |rec|
+        outf rec.to_s
+      end
+    end
+
+    def header
+      @decorations.join("\t")
+    end
+
+    # Runs the query and yields the records
+    def records
+      return enum_for(:records) unless block_given?
+
+      Cluster.where(restrictions_as_hash).no_timeout.to_a.each do |clusterable|
+        clusterable.send(@base).each do |embedded_doc|
+          # This is almost like a mongo aggregate,
+          # where the `Cluster.where(...)`    is the outer `$match`,
+          # the `clusterable.send(@base)` is the `$unwind: @base`,
+          # and the `matching_record?(embedded_doc)` is the inner `$match`.
+          # Almost. And perhaps that would have been the way to go.
+          next unless matching_record?(embedded_doc)
+          # If we passed matching_record?, then the restrictions are satisfied.
+          # Get the fields in @decoration from the embedded doc, for the output.
+          extracted_fields = @decorations.map { |d| embedded_doc.send(d.key) }
+
+          if extracted_fields.empty?
+            # We may be validating to the point where we can't end up here.
+            # Keeping it during development but TODO remove it.
+            raise ArgumentError, "No fields extractable from #{@base} document"
+          end
+
+          yield extracted_fields
+        end
+      end
+    end
+
+    def restrictions_as_hash
+      # from [ {a.b => c}, {d.e => f} ]
+      # to   {  a.b => c,   d.e => f  }
+      @restrictions.map do |r|
+        [r.on_key, r.val]
+      end.to_h
+    end
+
+    # Does the embedded doc under observation match all our @restrictions?
+    def matching_record?(embedded_doc)
+      matches = []
+      @restrictions.each do |restriction|
+        # If a restriction is {holdings.ocn => 5},
+        # then check the embedded doc (a holdings rec),
+        # and see if its restriction.key field (ocn)
+        # equals the restriction val (5)?
+        doc_field_val = embedded_doc.send(get_field(restriction.key))
+        # Add a truth value to array.
+        matches << (doc_field_val == restriction.val)
+      end
+      # For this embedded_doc, as we checked all restrictions,
+      # did we add only trues? If so, the embedded_doc is a match.
+      matches.all?(true)
+    end
+
+    # Ensure that the model class, e.g. Clusterable::XY, has the field Z
+    def validate_model_field(model_class, field)
+      unless Dynamic.model[model_class].fields.key?(field)
+        raise ArgumentError, "Model #{model_class} does not have field #{field}"
+      end
+    end
+
+    # Given a string, figure out model_class. ("x.y"->"x", "y"->@base)
+    def get_model_class(str)
+      if /\./.match?(str)
+        validate_clusterable(str.split(".").first)
+      else
+        @base
+      end
+    end
+
+    # Given a string, figure out field. ("x.y"->"y", "y"->"y")
+    def get_field(str)
+      if /\./.match?(str)
```

Comment:
Under what conditions would this be used? This branch isn't covered by the tests

Reply:
It *is* used. Once. Made it private.

See `doc_field_val = embedded_doc.send(get_field(restriction.key))`
... which sort of evaluates to:
`doc_field_val = commitment.organization`
... or whatever `clusterable.field` is currently being sought.